Feature/job groups (#24558)
cilium-team/cilium: Documentation/contributing/development/hive.rst

        }
    }

Job groups
^^^^^^^^^^

The `job package <https://pkg.go.dev/github.com/cilium/cilium/pkg/hive/job>`_ contains logic that 
makes it easy to manage units of work that the package refers to as "jobs". These jobs are 

> scheduled as part of a job group. These jobs themselves come in a variety of flavors.

Reviewer: For better localization:
```suggestion
scheduled as part of a job group. These jobs themselves come in several varieties.
```

> Every job, fundamentally is a callback function provided by the user with additional logic which
> is slightly different for each job type. The jobs and groups manage a lot of the boilerplate
> surrounding lifecycle management. The callbacks are called from the job to perform the actual
> work.
> 
> Take the following somewhat contrived example:

Reviewer: Grammar, clarity
```suggestion
Every job is a callback function provided by the user with additional logic which
differs slightly for each job type. The jobs and groups manage a lot of the boilerplate
surrounding lifecycle management. The callbacks are called from the job to perform the actual
work.

Consider the following example:
```

        }

        return nil
    }

    func (ex *exampleCell) daemon(ctx context.Context) error {
        for {
            randomTimeout := time.NewTimer(time.Duration(rand.Intn(3000)) * time.Millisecond)
            select {
            case <-ctx.Done():
                return nil

            case <-randomTimeout.C:
                ex.doSomeWork()
            }
        }
    }

    func (ex *exampleCell) timer(ctx context.Context) error {
        if err := ex.doSomeWork(); err != nil {
            return fmt.Errorf("doSomeWork: %w", err)
        }

        return nil
    }

    func (ex *exampleCell) Trigger() {
        ex.trigger.Trigger()
    }

    func (ex *exampleCell) observer(ctx context.Context, event struct{}) error {
        ex.logger.Info("Observed event")
        return nil
    }

    func (ex *exampleCell) HeavyLifting() {
        ex.jobGroup.Add(job.OneShot("long-running-job", func(ctx context.Context) error {
            for i := 0; i < 1_000_000; i++ {
                // Do some heavy lifting
            }
            return nil
        }))
    }

    func (ex *exampleCell) doSomeWork() error {
        ex.workChan <- struct{}{}
        return nil
    }

> The above example shows a number of use-cases in one cell. We start by requesting the job.Registry
> via the constructor. We can use the registry to create job groups, in most cases one will be enough.
> To this group we can add our jobs in the constructor. Any jobs added in the constructor are queued
> until the lifecycle of our cell starts. The group is added to the lifecycle and manages this 
> internally. Jobs can also be added at runtime which can be handy for dynamic workloads while still
> guaranteeing a clean shutdown.

Reviewer: Grammar, clarity, use active voice in present tense, avoid "we" and focus instead on a reader's actions ("you" often works)
```suggestion
The preceding example shows a number of use cases in one cell. The cell starts by requesting the job.Registry
by way of the constructor. The registry can create job groups; in most cases, one is enough.
You can add jobs in the constructor to this group. Any jobs added in the constructor are queued
until the lifecycle of the cell starts. The group is added to the lifecycle and manages jobs 
internally. You can also add jobs at runtime, which can be handy for dynamic workloads while still
guaranteeing a clean shutdown.
```

A job group will cancel the context to all jobs when the lifecycle ends. Any job callbacks are 
expected to exit as soon as possible when the ``ctx`` is "Done". The group will make sure that all
jobs are properly shutdown before the cell stops. If callbacks that do not stop within reasonable 

> amount of time may cause hive to perform a hard shutdown.

Reviewer: Use present tense, grammar
```suggestion
A job group cancels the context to all jobs when the lifecycle ends. Any job callbacks are 
expected to exit as soon as the ``ctx`` is "Done". The group makes sure that all
jobs are properly shut down before the cell stops. Callbacks that do not stop within a reasonable 
amount of time may cause the hive to perform a hard shutdown.

> There are 3 job types: one-shot jobs, timer jobs, and observer jobs. One shot jobs run a limited 
> amount of times, they can be used for short running jobs or jobs that span the entire lifecycle.
> Once the callback exits without error, its never called again. A one-shot can optionally have retry
> logic and/or trigger hive shutdown if it fails. Timers are called on a specified interval but they
> can also be externally triggered. Lastly, we have observer jobs which are invoked for every event
> on a ``stream.Observable``.

Reviewer: Consistent hyphens, grammar
```suggestion
There are 3 job types: one-shot jobs, timer jobs, and observer jobs. One-shot jobs run a limited 
number of times: you can use them for brief jobs, or for jobs that span the entire lifecycle.
Once the callback exits without error, it is never called again. Optionally, a one-shot job can include retry
logic and/or trigger hive shutdown if it fails. Timers are called on a specified interval but they
can also be externally triggered. Lastly, observer jobs are invoked for every event
on a ``stream.Observable``.
```

Reviewer 2: @zacharysarah thoughts on "use them" vs. "you can use them"? The former seems more concise and imperative.

(Otherwise all of this feedback looks good, @dylandreimerink / @zacharysarah do you think either of you would be able to apply these diffs in a fresh PR to the tree?)

Reviewer: @joestringer Sorry, I'm just now seeing this comment. You could use either, but "use them" is totally appropriate here, and I kinda like it.
